File: cloud-backend/internal/ai/circuit.go
package ai

import (
	"strings"
)

type CircuitClass struct {
	Code   string `json:"code"`
	Class  string `json:"class"`
	Reason string `json:"reason"`
}

type ModuleHit struct {
	Name      string   `json:"name"`
	TxID      string   `json:"tx_id"`
	RxID      string   `json:"rx_id"`
	Family    string   `json:"family,omitempty"`
	Confirmed bool     `json:"confirmed"`
	Reachable bool     `json:"reachable"`
	DTCs      []string `json:"dtcs,omitempty"`
	Error     string   `json:"error,omitempty"`
}

type NetworkHint struct {
	Reading string `json:"reading"`
	Summary string `json:"summary"`
	Live    int    `json:"live"`
	Dark    int    `json:"dark"`
}

func ClassifyCode(code, title string) CircuitClass {
	code = strings.ToUpper(strings.TrimSpace(code))
	title = strings.ToLower(strings.TrimSpace(title))
	out := CircuitClass{Code: code, Class: "component", Reason: "No circuit or bus pattern in the code or title."}
	if code == "" {
		return out
	}
	if code[0] == 'U' {
		if strings.Contains(code, "0073") || strings.Contains(title, "bus off") {
			return CircuitClass{Code: code, Class: "bus_off", Reason: "Network bus-off / U0073-class."}
		}
		return CircuitClass{Code: code, Class: "lost_communication", Reason: "U-code: lost communication with a module."}
	}
	switch {
	case hasAny(title, "short to battery", "short to batt", "short to b+", "circuit high"):
		return CircuitClass{Code: code, Class: "short_to_battery", Reason: "Title is circuit-high / short to battery."}
	case hasAny(title, "short to ground", "short to gnd", "circuit low"):
		return CircuitClass{Code: code, Class: "short_to_ground", Reason: "Title is circuit-low / short to ground."}
	case hasAny(title, "open circuit", "circuit open", "open in"):
		return CircuitClass{Code: code, Class: "open_circuit", Reason: "Title is an open circuit."}
	case strings.Contains(title, "circuit"):
		return CircuitClass{Code: code, Class: "circuit", Reason: "Title names a circuit, not only a component."}
	}
	return out
}

func ClassifyCodes(codes []string, titles map[string]string) []CircuitClass {
	out := make([]CircuitClass, 0, len(codes))
	for _, c := range codes {
		title := ""
		if titles != nil {
			title = titles[strings.ToUpper(strings.TrimSpace(c))]
		}
		out = append(out, ClassifyCode(c, title))
	}
	return out
}

func WiringShaped(classes []CircuitClass) bool {
	for _, c := range classes {
		switch c.Class {
		case "open_circuit", "short_to_battery", "short_to_ground", "circuit", "lost_communication", "bus_off":
			return true
		}
	}
	return false
}

func InferNetwork(mods []ModuleHit) NetworkHint {
	live, dark := 0, 0
	ecmUp := false
	var confirmedDark []string
	for _, m := range mods {
		if m.Reachable {
			live++
			if strings.EqualFold(m.Name, "ECM") {
				ecmUp = true
			}
			continue
		}
		dark++
		if m.Confirmed {
			confirmedDark = append(confirmedDark, m.Name)
		}
	}
	h := NetworkHint{Live: live, Dark: dark}
	if !ecmUp {
		h.Reading = "backbone"
		h.Summary = "ECM did not answer. Check DLC power, ground, and CAN before blaming a single module."
		return h
	}
	if len(confirmedDark) > 0 {
		h.Reading = "branch"
		h.Summary = "ECM is live. Silent confirmed node: " + strings.Join(confirmedDark, ", ") + ". That branch (power / ground / CAN), not the whole bus."
		return h
	}
	if dark >= 3 {
		h.Reading = "probes_silent"
		h.Summary = "ECM is live. Extra Toyota 11-bit probes did not answer — they may be absent on this car. Do not treat as a backbone failure."
		return h
	}
	h.Reading = "ok"
	h.Summary = "Probed modules answered."
	return h
}

func hasAny(s string, needles ...string) bool {
	for _, n := range needles {
		if strings.Contains(s, n) {
			return true
		}
	}
	return false
}

File: diagnostic-worker/mechazone_worker/circuit.py
"""Classify DTCs as circuit / bus vs component. Does not invent pin data."""

from __future__ import annotations

from typing import Any


def classify_code(code: str, title: str = "") -> dict[str, str]:
    code = (code or "").strip().upper()
    title = (title or "").strip().lower()
    out = {"code": code, "class": "component", "reason": "No circuit or bus pattern in the code or title."}
    if not code:
        return out
    if code.startswith("U"):
        if "0073" in code or "bus off" in title:
            return {"code": code, "class": "bus_off", "reason": "Network bus-off / U0073-class."}
        return {"code": code, "class": "lost_communication", "reason": "U-code: lost communication with a module."}
    if any(s in title for s in ("short to battery", "short to batt", "short to b+", "circuit high")):
        return {"code": code, "class": "short_to_battery", "reason": "Title is circuit-high / short to battery."}
    if any(s in title for s in ("short to ground", "short to gnd", "circuit low")):
        return {"code": code, "class": "short_to_ground", "reason": "Title is circuit-low / short to ground."}
    if any(s in title for s in ("open circuit", "circuit open", "open in")):
        return {"code": code, "class": "open_circuit", "reason": "Title is an open circuit."}
    if "circuit" in title:
        return {"code": code, "class": "circuit", "reason": "Title names a circuit, not only a component."}
    return out


def classify_codes(codes: list[str], titles: dict[str, str] | None = None) -> list[dict[str, str]]:
    titles = titles or {}
    return [classify_code(c, titles.get(c.upper(), titles.get(c, ""))) for c in codes]


def wiring_shaped(classes: list[dict[str, str]]) -> bool:
    return any(
        row["class"] in {
            "open_circuit",
            "short_to_battery",
            "short_to_ground",
            "circuit",
            "lost_communication",
            "bus_off",
        }
        for row in classes
    )


def network_hint(modules: list[dict[str, Any]]) -> dict[str, Any]:
    live = [m for m in modules if m.get("reachable")]
    dark = [m for m in modules if not m.get("reachable")]
    ecm_up = any(m.get("name") == "ECM" and m.get("reachable") for m in live)
    confirmed_dark = [str(m.get("name")) for m in dark if m.get("confirmed")]
    hint: dict[str, Any] = {"live": len(live), "dark": len(dark)}
    if not ecm_up:
        hint["reading"] = "backbone"
        hint["summary"] = "ECM did not answer. Check DLC power, ground, and CAN before blaming a single module."
        return hint
    if confirmed_dark:
        hint["reading"] = "branch"
        hint["summary"] = (
            "ECM is live. Silent confirmed node: "
            + ", ".join(confirmed_dark)
            + ". That branch (power / ground / CAN), not the whole bus."
        )
        return hint
    if len(dark) >= 3:
        hint["reading"] = "probes_silent"
        hint["summary"] = (
            "ECM is live. Extra Toyota 11-bit probes did not answer — they may be absent on this car. "
            "Do not treat as a backbone failure."
        )
        return hint
    hint["reading"] = "ok"
    hint["summary"] = "Probed modules answered."
    return hint
